packages/cli/src/ui/components/MainContent.tsx

```diff
@@ -11,19 +11,23 @@ import { OverflowProvider } from '../contexts/OverflowContext.js';
 import { useUIState } from '../contexts/UIStateContext.js';
 import { useAppContext } from '../contexts/AppContext.js';
 import { AppHeader } from './AppHeader.js';
-import { useSettings } from '../contexts/SettingsContext.js';
+import { useAlternateBuffer } from '../hooks/useAlternateBuffer.js';
+import { SCROLL_TO_ITEM_END } from './shared/VirtualizedList.js';
+import { ScrollableList } from './shared/ScrollableList.js';
+import { useMemo, memo, useCallback } from 'react';
+import { MAX_GEMINI_MESSAGE_LINES } from '../constants.js';
+
+const MemoizedHistoryItemDisplay = memo(HistoryItemDisplay);
+const MemoizedAppHeader = memo(AppHeader);
 
 // Limit Gemini messages to a very high number of lines to mitigate performance
 // issues in the worst case if we somehow get an enormous response from Gemini.
 // This threshold is arbitrary but should be high enough to never impact normal
 // usage.
-const MAX_GEMINI_MESSAGE_LINES = 65536;
-
 export const MainContent = () => {
```

Comment:
Nit: might be worth refactoring maincontent.tsx, it's becoming complex because it handles two different rendering strategies: the legacy Static rendering (for standard buffer) and the new ScrollableList virtualization (for alternate buffer). We could split MainContent into smaller, focused components. This isolates the logic and makes the main entry point cleaner.

this could also be a good follow up, so we don't increase the size of this PR